File: madgraph/iolibs/template_files/subtraction/subtraction_schemes/torino/variables.py
##########################################################################################
#
# Copyright (c) 2009 The MadGraph5_aMC@NLO Development team and Contributors
#
# This file is a part of the MadGraph5_aMC@NLO project, an application which 
# automatically generates Feynman diagrams and matrix elements for arbitrary
# high-energy processes in the Standard Model and beyond.
#
# It is subject to the MadGraph5_aMC@NLO license which should accompany this 
# distribution.
#
# For more information, visit madgraph.phys.ucl.ac.be and amcatnlo.web.cern.ch
#
##########################################################################################

import os
import math

import madgraph.integrator.vectors as vectors
import madgraph.various.misc as misc
import madgraph.integrator.mappings as mappings
import factors_and_cuts as fc

from madgraph.integrator.vectors import Vector, LorentzVector

import commons.utils as utils

import logging



logger = logging.getLogger('madgraph')


CurrentImplementationError = utils.CurrentImplementationError


def get_sudakov_decomp_FF(ki, kj, kr):
    """returns the sudakov decomposition (x and kt) of ki and kj, with kr being the other
    reference momentum.
    Follows eq 2.24-2.27 of the torino paper"""

    sij = ki.dot(kj)
    sir = ki.dot(kr)
    sjr = kj.dot(kr)

    k = ki + kj
    kbar = k - sij / (sir + sjr) * kr

    xi = sir / (sir + sjr)
    #xj = sjr / (sir + sjr)
    xj = 1. - xi 

    kitil = ki - k * xi  - kr * (k.dot(ki) / k.square() - xi) * k.square() / k.dot(kr)
    kjtil = kj - k * xj  - kr * (k.dot(kj) / k.square() - xj) * k.square() / k.dot(kr)

    y = sij / (sij + sir + sjr)    #for final kr
    x = 1. - sij / (sir + sjr)     #for initial kr

    return xi, xj, kitil, kjtil, y, x

def TRN_FFn_variables(higher_PS_point, qC, children, **opts):
    """
    returns variables for the torino subtraction scheme, in particular x's and
    ktil as returned from get_sudakov_decomp

    """

    all_p_fs = [higher_PS_point[child] for child in children]
    Q = opts['Q']
    # the recoiler
    p_rec = higher_PS_point[opts['ids']['c']]

    # Add additional ss's
    ss_i_j = {}
    for i_fs in range(len(all_p_fs)):
        for j_fs in range(i_fs+1,len(all_p_fs)):
            ss_i_j[(i_fs,j_fs)] = 2.*all_p_fs[i_fs].dot(all_p_fs[j_fs])
    ss_i_j_tot = sum(ss_i_j.values())
    for k,v in ss_i_j.items():
        ss_i_j[(k[1],k[0])] = v

    # now x's and kts
    xi, xj, kitil, kjtil, y, x = get_sudakov_decomp_FF(all_p_fs[0], all_p_fs[1], p_rec)

    xis = (xi, xj)
    kTs = (kitil, kjtil)

    return [{'xs':xis, 'kTs':kTs, 'ss':ss_i_j, 'y':y, 'x':x},]

def get_sudakov_decomp_IF(kj, ki, kr, kr_leg_number):
    """returns the sudakov decomposition (x and kt) of ki and kj, with kr being the other
    reference momentum for the IF case. kj=initial, ki=final 
    Follows eq 1.18-1.2 of the nlo_v4 note"""

    sij = ki.dot(kj)
    sir = ki.dot(kr)
    sjr = kj.dot(kr)

    na, nb = mappings.InitialCollinearVariables.collinear_and_reference(kj)
    pa = kj
        # Compute the sum of momenta
    pA = LorentzVector(pa)
    pA -= ki
    # Pre-compute variables
    napA = na.dot(pA)
    nbpA = nb.dot(pA)
    nanb = na.dot(nb)
    nbpa = nb.dot(pa)
    # zA = nbpA / nbpa
    ktA = pA - (nbpA * na + napA * nb) / nanb
    # Compute all kinematic variables
    pi = ki
    napi = na.dot(pi)
    nbpi = nb.dot(pi)
    # zi = nbpi / nbpa
    kti = pi - (nbpi*na+napi*nb) / nanb
    # kts
    kjtil = ktA
    kitil = kti

    if kr_leg_number > 2:
        xj = 1. - sir / (sij+sjr)
        xi = 1. - xj
    else:
        xj = 1. - ((sij + sir) / sjr)
        xi = 1. - xj

    # print('VAR - xj : ' + str(xj))
    # print('VAR - xi : ' + str(xi))

    # Test on kt
    # print('VAR - kjtil     : ' + str(kjtil))
    # print('VAR - test kt   : ' + str(ki-kj*(sir/sjr)-kr*(sij/sjr)))
    # print('VAR - test kt 2 : ' + str(kjtil.dot(kr)))
    # print('VAR - test kt 3 : ' + str(kjtil.dot(kj)))


    # Variable for initial splitting - final rec
    z = sij / (sij + sjr)
    # Variables for initial splitting - initial rec
    v = sij / (sij + sir)

    # x_r_final = 1. - (sjr / (sij + sir))
    # x_r_initial = 1. - ((sij + sjr) / sir)
    # #x1 = 1. - (sjr / (sir - sij))    #for soft initial case
    # x1 = 1. - ((sij + sjr) / sir)    #for soft initial case

    
    return xj, xi, kjtil, kitil, z, v

    # return xj, xi, kjtil, kitil, x_r_final, x_r_initial, z, v, x1

def TRN_IFn_variables(higher_PS_point, qC, children, **opts):
    """
    returns variables for the torino subtraction scheme, in particular x's and
    ktil as returned from get_sudakov_decomp_IF

    """

    all_p_fs = [higher_PS_point[child] for child in children]
    Q = opts['Q']
    # the recoiler
    p_rec = higher_PS_point[opts['ids']['c']]
    p_rec_leg_number = opts['ids']['c']

    # print('VAR : all_p_fs : ' + str(all_p_fs))
    # print('VAR : Q        : ' + str(Q))
    # print('VAR : p_rec    : ' + str(p_rec))

    # Add additional ss's
    ss_i_j = {}
    for i_fs in range(len(all_p_fs)):
        for j_fs in range(i_fs+1,len(all_p_fs)):
            ss_i_j[(i_fs,j_fs)] = 2.*all_p_fs[i_fs].dot(all_p_fs[j_fs])
    ss_i_j_tot = sum(ss_i_j.values())
    for k,vv in ss_i_j.items():
        ss_i_j[(k[1],k[0])] = vv

    # now x's and kts where all_p_fs[0] = IN , all_p_fs[1] = F
    # xi, xj, kitil, kjtil, x_r_final, x_r_initial, z, v, x1 = get_sudakov_decomp_IF(all_p_fs[0], all_p_fs[1], p_rec, p_rec_leg_number)

    # xis = (xi, xj)
    # kTs = (kitil, kjtil)
    # x_mapping = (x_r_final, x_r_initial)
    # return [{'xs':xis, 'kTs':kTs, 'ss':ss_i_j, 'x_mapping':x_mapping, 'z':z, 'v':v, 'x1':x1},]

    xj, xi, kjtil, kitil, z, v = get_sudakov_decomp_IF(all_p_fs[0], all_p_fs[1], p_rec, p_rec_leg_number)

    xis = (xj, xi)
    kTs = (kjtil, kitil)
 
    return [{'xs':xis, 'kTs':kTs, 'ss':ss_i_j, 'z':z, 'v':v},]

File: tmp_fortran/variables.f
      subroutine trn_ffn_variables(pp,children,xis,kts,ss_ij,y,x)
      implicit none
      include 'nexternal.inc'
      integer i,j
      real * 8 pp(0:3,NEXTERNAL),allpfs(0:3,2)
      integer children(2)
      real * 8 ss_ij(1),ss_ijtot
      real * 8 xi,xj,kitil(0:3),kjtil(0:3),y,x
      real * 8 xis(2),kts(0:3,2)
      real * 8 prec(0:3)
      real * 8 dot
      integer rec_legnumber
      
      ss_ijtot=0d0

      !!!!!! This only works for NLO computation with 3 particles 
      !!!!!! in the final state.
      
      
      do i=1,2
         allpfs(:,i) = pp(:,children(i))
      enddo

!     TODO
!     call get_recoiler(...,prec,rec_legnumber)



      ss_ij(1) = 2d0*dot(allpfs(:,1),allpfs(:,2))
      
      do i=1,1 !!!!! TO BE GENERALIZED LATER
         ss_ijtot = ss_ijtot + ss_ij(i)
      enddo


      call GET_SUDAKOV_DECOMP_FF(allpfs(:,1),allpfs(:,2),prec,xi,xj,
     $     kitil,kjtil,y,x)
      
      xis = [xi,xj]
      kts(:,1) = kitil
      kts(:,2) = kjtil
      
      
      end




      subroutine trn_ifn_variables(pp,children,xis,kts,ss_ij,z,v)
      implicit none
      include 'nexternal.inc'
      integer i,j
      real * 8 pp(0:3,NEXTERNAL),allpfs(0:3,2)
      integer children(2)
      real * 8 ss_ij(1),ss_ijtot
      real * 8 xi,xj,kitil(0:3),kjtil(0:3),z,v,y,x
      real * 8 xis(2),kts(0:3,2)
      real * 8 prec(0:3)
      integer rec_legnumber
      real * 8 dot

      ss_ijtot=0d0

      !!!!!! This only works for NLO computation with 3 particles 
      !!!!!! in the final state.
      
      
      do i=1,2
         allpfs(:,i) = pp(:,children(i))
      enddo

!     TODO
!     call get_recoiler(...,prec,rec_legnumber)



      ss_ij(1) = 2d0*dot(allpfs(:,1),allpfs(:,2))
      
      do i=1,1 !!!!! TO BE GENERALIZED LATER
         ss_ijtot = ss_ijtot + ss_ij(i)
      enddo


      call GET_SUDAKOV_DECOMP_IF(allpfs(:,1),allpfs(:,2),prec,
     $     rec_legnumber,xi,xj,kitil,kjtil,y,x)
      
      xis = [xi,xj]
      kts(:,1) = kitil
      kts(:,2) = kjtil
      
      
      end



      SUBROUTINE GET_SUDAKOV_DECOMP_FF(ki,kj,kr,xi,xj,kitil,kjtil,y,x)
      IMPLICIT NONE
C
C---- returns the sudakov decomposition (x and kt) of ki and kj, with kr being the other
C---- reference momentum
C
C     
C     ARGUMENTS 
C
      double precision ki(0:3),kj(0:3),kr(0:3),kitil(0:3),kjtil(0:3)
      double precision xi,xj,y,x
CF2PY INTENT(IN)  :: ki
CF2PY INTENT(IN)  :: kj
CF2PY INTENT(IN)  :: kr
CF2PY INTENT(OUT) :: xi
CF2PY INTENT(OUT) :: xj
CF2PY INTENT(OUT) :: kitil
CF2PY INTENT(OUT) :: kjtil
CF2PY INTENT(OUT) :: y
CF2PY INTENT(OUT) :: x

C     LOCAL
      double precision sij,sir,sjr,k2
      double precision k(0:3)
C     
C     EXTERNAL
C     
      real * 8 dot
C      EXTERNAL dot


      sij=dot(ki,kj)
      sir=dot(ki,kr)
      sjr=dot(kj,kr)

      k=ki(:)+kj(:)
      k2=dot(k,k)
      xi=sir/(sir+sjr)
      xj=1d0-xi

      kitil=ki-k*xi-kr*((dot(k,ki)/k2)-xi)*(k2/dot(k,kr))
      kjtil=kj-k*xj-kr*((dot(k,kj)/k2)-xj)*(k2/dot(k,kr))

C---- for final kr
      y=sij/(sij + sir + sjr)
C---- for initial kr
      x=1d0-sij/(sir + sjr)

      END


      SUBROUTINE GET_SUDAKOV_DECOMP_IF(kj,ki,kr,kr_leg_number,
     $      xj,xi,kjtil,kitil,z,v)
      IMPLICIT NONE
C
C---- returns the sudakov decomposition (x and kt) of ki and kj, with kr being the other
C---- reference momentum for the IF case. kj=initial, ki=final
C
C     
C     ARGUMENTS 
C
      integer kr_leg_number
      double precision kj(0:3),ki(0:3),kr(0:3),kitil(0:3),kjtil(0:3)
      double precision xi,xj,z,v
CF2PY INTENT(IN)  :: ki
CF2PY INTENT(IN)  :: kj
CF2PY INTENT(IN)  :: kr
CF2PY INTENT(IN)  :: kr_leg_number
CF2PY INTENT(OUT) :: xi
CF2PY INTENT(OUT) :: xj
CF2PY INTENT(OUT) :: kitil
CF2PY INTENT(OUT) :: kjtil
CF2PY INTENT(OUT) :: z
CF2PY INTENT(OUT) :: v
C
C     LOCAL
C
      double precision sij,sir,sjr
      double precision pa(0:3),pAA(0:3),ktA(0:3),kti(0:3),pi(0:3)
      real*8 napAA, nbpAA, nanb, nbpa, napi, nbpi
C     
C     EXTERNAL
C     
      real * 8 dot
c      EXTERNAL dot
      double precision na(0:3), nb(0:3)



      sij=dot(ki,kj)
      sir=dot(ki,kr)
      sjr=dot(kj,kr)

      call collinear_and_reference(kj,na,nb)

      pa = kj
C---- Compute the sum of momenta
      pAA = pa - ki
C---- Pre-compute variables
      napAA = dot(na,pAA)
      nbpAA = dot(nb,pAA)
      nanb = dot(na,nb)
      nbpa = dot(nb,pa)
C---- zA = nbpA / nbpa
      ktA = pAA - (nbpAA * na + napAA * nb) / nanb
C---- Compute all kinematic variables
      pi = ki
      napi = dot(na,pi)
      nbpi = dot(nb,pi)
C---- zi = nbpi / nbpa
      kti = pi - (nbpi*na + napi*nb) / nanb
C---- kts
      kjtil = ktA
      kitil = kti
  
      IF (kr_leg_number > 2) THEN
          xj = 1d0 - sir / (sij+sjr)
          xi = 1d0 - xj
      ELSE
          xj = 1d0 - ((sij + sir) / sjr)
          xi = 1d0 - xj
      END IF
  
C---- Variable for initial splitting - final rec
      z = sij / (sij + sjr)
C---- Variables for initial splitting - initial rec
      v = sij / (sij + sir)

      END




      SUBROUTINE collinear_and_reference(p,na,nb)
      IMPLICIT NONE
C****************************************************************************
C     Given a momentum, return normalized vectors on the light-cone
C****************************************************************************
      integer I
      double precision p(0:3)
      double precision n(0:2),na(0:3),nb(0:3)
CF2PY INTENT(IN)  :: p
CF2PY INTENT(OUT) :: na
CF2PY INTENT(OUT) :: nb

      !!!! This subroutine takes initial state momenta as input
      !!!! They are taken in the center of mass frame so they 
      !!!! do not have a transverse compoment wrt to the collision axis


      if(dsqrt(p(1)**2+p(2)**2).gt.1d-8) then
         write(*,*) 'The initial state momenta cannot have '
         write(*,*) 'a non-vanishing transverse momentum!'
         write(*,*) 'EXIT...'
         call exit(-1)
      endif

      n(0:1) = p(1:2) 
      n(2) = p(3)/abs(p(3))
      

C---- For good phase space points, p(0) >= 0, but test PS points might not be valid
      IF (p(0).ge.0d0) THEN
          na = (/1d0, n(0), n(1), n(2)/)
          nb = (/1d0, -n(0), -n(1), -n(2)/)
      ELSE
          na = (/1d0, -n(0), -n(1), -n(2)/)
          nb = (/1d0, n(0), n(1), n(2)/)
      END IF

      END
